### ISSUE TITLE
Whenever a new proxy is disconnected/connected, all other proxy seem to be disconnect

### ISSUE THREAD

**Issue by mshi** [rethinkdb/rethinkdb#4934]
Currently I have 4 servers clustered together. Each server instance that needs to access the database has a rethinkdb proxy service running. I have recently noticed that when one of the proxy server has a problem, and then I restart the service, it takes down all of the other proxy servers on _another_ instance with them. As a result, all the servers that are running the proxy will return errors along the lines of primary replica not available.

This is what I am starting the proxy with: `rethinkdb proxy --daemon --runuser rethinkdb --rungroup rethinkdb --pid-file /var/run/rethinkdb/instance1/proxy.pid --join 10.240.0.5:29015 -o 1 --log-file /var/lib/rethinkdb/instance1/data/log_file` And this is triggered in the init.d startup.

I am running 2.1.4 currently.

**Comment by danielmewes** [comment]
Hi @mshi, thanks for the report.
Could you post the log file for one of the proxies that gets disconnected?

**Comment by mshi** [comment]
Here is the log file for one of them. 
```
2015-10-09T00:46:53.176725986 22062.276531s notice: Disconnected from server "us_east_1" 0efb01f9-bfc5-48de-82d7-4689e9681ce3
2015-10-09T00:46:53.584416968 96457.418379s notice: Disconnected from server "us_east_1" 0efb01f9-bfc5-48de-82d7-4689e9681ce3
2015-10-09T00:53:04.806697760 96828.640660s notice: Disconnected from proxy 9b309d0f-bcae-4ee5-a3fd-99fadf5dd879
2015-10-09T00:53:05.916207448 96829.750170s notice: Connected to proxy 24c9e83d-f458-42db-9ec6-dee3f15fcee2
2015-10-09T00:59:59.472927905 97243.306890s notice: Connected to server "us_east_1" 0efb01f9-bfc5-48de-82d7-4689e9681ce3
2015-10-09T01:00:05.237927840 97249.071890s notice: Disconnected from proxy 72a2d778-9cee-4b36-81ee-faadaaa22125
2015-10-09T01:00:05.548777977 97249.382740s notice: Connected to proxy e1f41d87-fb01-4d83-8724-fbfc356c36ff
2015-10-09T01:00:06.651434746 22855.751240s notice: Connected to server "us_east_1" 0efb01f9-bfc5-48de-82d7-4689e9681ce3
```

Here is the log file of another that I think might be the suspect of something:
```
2015-10-09T00:46:51.629809588 21284.865798s error: Heartbeat timeout, killing connection to peer 10.240.0.3:29015
2015-10-09T00:46:51.630793034 21284.866781s notice: Disconnected from server "us_east_1" 0efb01f9-bfc5-48de-82d7-4689e9681ce3
2015-10-09T00:46:53.180192988 21286.416181s warn: Received invalid clustering header from 10.240.0.3:29015, closing connection -- something might be connecting to the wrong port.
2015-10-09T00:46:53.185767803 86569.987021s notice: Disconnected from server "us_east_1" 0efb01f9-bfc5-48de-82d7-4689e9681ce3
2015-10-09T00:53:04.533424915 21657.769413s notice: Server got SIGTERM from pid 18720, uid 0; shutting down...
2015-10-09T00:53:04.533594617 21657.769583s notice: Shutting down client connections...
2015-10-09T00:53:04.725465826 21657.961454s notice: All client connections closed.
2015-10-09T00:53:04.725493373 21657.961481s notice: Shutting down storage engine... (This may take a while if you had a lot of unflushed data in the writeback cache.)
2015-10-09T00:53:04.772368127 21658.008356s notice: Storage engine shut down.
2015-10-09T00:53:04.775600173 21658.011589s notice: Disconnected from server "us_east_2" 72947562-1a24-4e69-be06-d8426ba75d8c
2015-10-09T00:53:04.775702981 21658.011691s notice: Disconnected from server "cloud_us_primary_db" 33dbdd76-2f08-4b1f-9fa0-f4d39c476a80
2015-10-09T00:53:04.775710488 21658.011698s notice: Disconnected from server "cloud_us_primary" cf8f0469-d329-4473-9ff2-1485a6110814
2015-10-09T00:53:04.777948809 86941.579202s notice: Disconnected from proxy 9b309d0f-bcae-4ee5-a3fd-99fadf5dd879
2015-10-09T00:53:05.767185046 0.213481s notice: Listening for intracluster connections on port 29016
2015-10-09T00:53:05.767320151 0.213615s info: Attempting connection to 1 peer...
2015-10-09T00:53:05.770197182 0.216494s notice: Connected to server "cloud_us_primary_db" 33dbdd76-2f08-4b1f-9fa0-f4d39c476a80
2015-10-09T00:53:05.773690687 0.219986s notice: Listening for client driver connections on port 28016
2015-10-09T00:53:05.773705842 0.220001s notice: Listening for administrative HTTP connections on port 8081
2015-10-09T00:53:05.773707210 0.220002s notice: Listening on address: 127.0.0.1
2015-10-09T00:53:05.773708414 0.220003s notice: To fully expose RethinkDB on the network, bind to all addresses by running rethinkdb with the `--bind all` command line option.
2015-10-09T00:53:05.773709593 0.220005s notice: Proxy ready
2015-10-09T00:53:05.774245750 86942.575498s notice: Connected to proxy 24c9e83d-f458-42db-9ec6-dee3f15fcee2
2015-10-09T00:53:05.933443956 0.379739s notice: Connected to server "us_east_2" 72947562-1a24-4e69-be06-d8426ba75d8c
2015-10-09T00:53:05.934008960 0.380304s notice: Connected to server "cloud_us_primary" cf8f0469-d329-4473-9ff2-1485a6110814
2015-10-09T00:59:59.516923637 87356.318176s notice: Connected to server "us_east_1" 0efb01f9-bfc5-48de-82d7-4689e9681ce3
2015-10-09T01:00:05.255008919 87362.056262s notice: Disconnected from proxy 72a2d778-9cee-4b36-81ee-faadaaa22125
2015-10-09T01:00:05.660478119 87362.461731s notice: Connected to proxy e1f41d87-fb01-4d83-8724-fbfc356c36ff
```
The SIGTERM is probably from me restarting the proxy or server (i forgot)

**Comment by danielmewes** [comment]
That's really odd. When one of the proxies on a different node goes down, do you have to restart it manually or does it recover by itself eventually?

I'm a bit confused by some lines in the log file. Is it possible that you are using the same `log_file` for both the actual server and a proxy? I recommend using separate log files to make debugging easier and to avoid corrupting the log in case both processes try to write to it at the same time. Though that's most likely unrelated to the issue at hand.

**Comment by mshi** [comment]
Yes, sorry. I am using the same log_file for actual server and proxy. This is because I wanted the logs to update on the admin console.

And to answer your question, I have to restart it manually, or else it will continue to say primary replica not found in my application code.

Something unrelated, but I have 4 replicas of a table, but when the primary replica goes down, it doesn't seem to do the failover recovery by electing a new primary replica (or at least not in a quick time) and causes all my application code to fail. It also crashes the changefeed (although I think I saw something about this in the bug fixes for 2.1.5?)

**Comment by danielmewes** [comment]
One possible source of issues could be that proxies right now appear to try connecting to each other.
If all of the proxies only bind to `localhost`, different proxies might try to establish connections to `localhost` to reach some of the other proxies. However there they won't find the expected other proxies, but will end up connecting to themselves of course.
I don't see any indication of that happening in the log file, but it's the only thing I can think of that could maybe cause some issues.

Could you try adding the following argument to the proxies?
`--canonical-address 1.1.1.1`
You can alternatively put that in the proxys' instance configurations like so:
`canonical-address=1.1.1.1`

(I'm using 1.1.1.1 here as an address that just doesn't exist. I hope that's true. At least it doesn't appear to respond to pings)

This should stop proxies from trying to connect to each other, which might or might not help.

If it doesn't we'll have to investigate this further.

**Comment by danielmewes** [comment]
The failover not happening is also weird. Can all servers connect to each other or could there be some firewall restrictions between them that make connections only work in one direction?

Changefeeds currently always get interrupted if the primary fails. 2.1.5 fixed a race condition that sometimes made it impossible to create new changefeeds even *after* the primary had come back up (or a failover was performed).

**Comment by mshi** [comment]
I'm running all of these on compute engine and I've set every canonical address to their local address for each config that host the database. And all internal IPs have all ports unblocked, so I don't think that's the case. I see them all connected in the admin panel with no issues reported. How long does the failover recovery take approximately?

Would the appropriate approach be to catch the error and then continuously retry or is there a better way of handling this?

I will try to set the canonical address of the proxy instances and give it a try later today during non-peak hours.

Thanks!

**Comment by danielmewes** [comment]
Failover should usually take in the range of 15 seconds.
If you could send me a copy of the log_file from the different servers for the time where you had the primary down, that might be helpful to figure out where things went wrong. My email is daniel@rethinkdb.com .

During that period queries will fail, so retrying on `ReqlOpFailedError` exceptions is often the right approach. You can add in a `r.db(...).wait()` into the loop, which avoids busy-waiting (it will only return when the next operation on that database is expected to succeed again).

**Comment by hamiltop** [comment]
I've just noticed a bunch of different issues with proxy that may or may not be related. Here's a big one:

1. Bring up cluster of 3 servers.
2. Set one table to 3 shards with 3 replicas.
3. Start proxy on local node. Proxy connects to all servers.
4. Terminate one server. Rebuild server.
5. Reconfigure table to 3 shards with 3 replicas.
6. Try to query data whose primary shard is on new server.

It seems the proxy is never made aware of the new server once the table is reconfigured. It throws `"Cannot perform read: primary replica for shard [\"SaY\\x7D\\x01\\x9D\", \"Scc\\x7E\\xA2F\") not available"`

This might even be reproducible by simply adding a new shard.

Worth filing a new ticket? or related to this?

**Comment by larkost** [comment]
@hamiltop: I will write up a test for the clear case you described immediately above. If you want to open a bug for that case it would be nice, otherwise if I complete the test first I will make one (I use the bug number in naming the test). We already have a few similar tests, but not one straight on the nose for that.

**Comment by larkost** [comment]
@hamiltop: was that last `Cannot perform read:` for just a moment, or did it last? If it was done immediately after either the server coming back up, or after the reconfigure then it would be expected: there is a brief moment during handover where there is no primary for a shard. We do this to guarantee that there are never two (and the incumbent possibilities of data corruption/loss). There is [an issue](https://github.com/rethinkdb/rethinkdb/issues/4331) open already to auto-retry queries that fall in this hole, but it has not yet been scheduled.

**Comment by hamiltop** [comment]
Looking at it again, it might have been due to network acls. The proxy was
not reachable from the other nodes. It's probably still a bug, since it
should fail more explicitly.

The error lasted until I restarted the proxy.

**Comment by larkost** [comment]
@hamiltop: Ok... that explains why I have not been able to reproduce it. We do have an issue for non-transitive connectivity: #4936. So I think we have this later part covered.

**Comment by hamiltop** [comment]
Great!

**Comment by danielmewes** [comment]
The fact that our proxies currently want to be reachable from the other cluster nodes is very annoying and I think we should definitely change that.

Though I think that even if the proxy isn't reachable from the other server, the proxy should still reach out to the other server on its own and be able to connect to it.
@hamiltop Is there a chance that the network configuration blocked connections in both directions, i.e. also from the proxy to the other server?

If not we definitely need to investigate this more.

**Comment by danielmewes** [comment]
We're changing the reconnection logic for proxies in order to fix https://github.com/rethinkdb/rethinkdb/issues/5456 . In the future, servers (including proxies) won't try to connect to a proxy more than once. Instead the proxy itself will be responsible for reconnecting to all servers once it loses a connection.

This might resolve this issue as well, though I'm not sure since we have never been able to reproduce it.

**Comment by hamiltop** [comment]
I'm happy to try to reproduce it once the fix is in.
